### PR TITLE
NXP-29984: Runtime Registries Migrations #4

### PR DESCRIPTION
**Pull request by atchertchian**

### CODE REVIEW

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XValueFactory.java [hunk 1]
```diff
@@ -196,6 +197,9 @@ public String serialize(Context context, Object value) {
     public static final XValueFactory CLASS = new XValueFactory() {
         @Override
         public Object deserialize(Context context, String value) {
+            if (StringUtils.isBlank(value)) {
```
Comment (atchertchian) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/XValueFactory.java:200]: This change was needed to allow disabling a query maker without redefining the class: see https://jira.nuxeo.com/browse/NXP-30092 for proper handling of such cases, with addition of unit tests.

**Comment by atchertchian** [comment]
(force-pushed after merge of previous PR #4593 on master)

File: modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/thumbnail/ThumbnailServiceImpl.java [hunk 2]
```diff
@@ -41,38 +38,39 @@
...
 public class ThumbnailServiceImpl extends DefaultComponent implements ThumbnailService {
 
-    private static final Log log = LogFactory.getLog(BlobHolderAdapterComponent.class);
-
     public static final String THUMBNAILFACTORY_EP = "thumbnailFactory";
 
     protected ThumbnailFactory defaultFactory;
 
-    protected final Map<String, ThumbnailFactory> factoriesByDocType = new HashMap<>();
+    protected Map<String, ThumbnailFactory> factoriesByDocType = new HashMap<>();
```
Comment (efge) [inline modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/thumbnail/ThumbnailServiceImpl.java:45]: Can be `null` given that you set them in `start`

File: modules/core/nuxeo-core-api/src/test/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentServiceTest.java [hunk 3]
```diff
@@ -91,7 +92,7 @@ public void testContrib() throws Exception {
     @Test
     @Deploy("org.nuxeo.ecm.core.api.tests:OSGI-INF/test-pathsegment-contrib.xml")
     public void testContribOverride() throws Exception {
-        PathSegmentService service = Framework.getService(PathSegmentService.class);
+        assertNotNull(service);
```
Comment (efge) [inline modules/core/nuxeo-core-api/src/test/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentServiceTest.java:95]: Not needed, test framework will throw if it cannot inject. Same in other methods.

File: modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java [hunk 4]
```diff
@@ -18,135 +18,53 @@
...
+    public void start(ComponentContext context) {
+        PathSegmentService confService = this.<PathSegmentServiceDescriptor> getRegistryContribution(XP).map(desc -> {
+            try {
+                return desc.className.getDeclaredConstructor().newInstance();
+            } catch (ReflectiveOperationException e) {
+                throw new RuntimeServiceException(e);
+            }
+        }).orElse(null);
+        if (confService == null) {
+            confService = new PathSegmentServiceDefault();
```
Comment (efge) [inline modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java:45]: Coud be predefined in a constant, and in the `orElse`

File: modules/core/nuxeo-core-persistence/src/main/java/org/nuxeo/ecm/core/persistence/PersistenceComponent.java [hunk 5]
```diff
@@ -100,7 +60,7 @@ public PersistenceProvider newProvider(String name) {
 
     @Override
     public HibernateConfiguration getHibernateConfiguration(String name) {
-        HibernateConfiguration config = registry.get(name);
+        HibernateConfiguration config = this.<HibernateConfiguration> getRegistryContribution(XP, name).orElse(null);
         if (config == null) {
             throw new PersistenceError("no hibernate configuration identified by '" + name + "' is available");
```
Comment (efge) [inline modules/core/nuxeo-core-persistence/src/main/java/org/nuxeo/ecm/core/persistence/PersistenceComponent.java:65]: `orElseThrow`?

File: modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/jdbc/QueryMakerServiceImpl.java [hunk 6]
```diff
@@ -20,88 +20,43 @@
...
-        }
-    }
-
-    @Override
-    public void unregisterContribution(Object contribution, String extensionPoint, ComponentInstance contributor) {
-        if (XP.equals(extensionPoint)) {
-            unregisterQueryMaker((QueryMakerDescriptor) contribution);
+    public void start(ComponentContext context) {
+        queryMakers = new ArrayList<>(2);
+        queryMakers.addAll(this.<QueryMakerDescriptor> getRegistryContributions(XP)
```
Comment (efge) [inline modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/main/java/org/nuxeo/ecm/core/storage/sql/jdbc/QueryMakerServiceImpl.java:42]: Could probably assign directly. `Collectors.toList` in practice returns an `ArrayList` and this won't change (as too many people already rely on that, per OpenJDK dev discussions).

File: modules/core/nuxeo-core-storage/src/main/java/org/nuxeo/ecm/core/storage/lock/LockManagerService.java [hunk 7]
```diff
@@ -41,109 +38,48 @@
...
+    public void start(ComponentContext context) {
+        lockManagers = new ConcurrentHashMap<>();
+        this.<LockManagerDescriptor> getRegistryContributions(XP_LOCKMANAGER).forEach(desc -> {
+            if (desc.klass != null) {
+                try {
+                    Constructor<? extends LockManager> ctor = desc.klass.getConstructor(String.class);
+                    lockManagers.put(desc.name, ctor.newInstance(name));
+                } catch (ReflectiveOperationException e) {
+                    addRuntimeMessage(Level.ERROR, e.getMessage());
+                    throw new NuxeoException(e);
```
Comment (efge) [inline modules/core/nuxeo-core-storage/src/main/java/org/nuxeo/ecm/core/storage/lock/LockManagerService.java:63]: `RuntimeServiceException` for consistency?

File: modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java (hunk 4 above)
Comment (atchertchian) [reply modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java:45]: Sadly using orElse directly does not work here:
```
The method orElse(capture#1-of ? extends PathSegmentService) in the type Optional<capture#1-of ? extends PathSegmentService> is not applicable for the arguments (PathSegmentService)
```

I'll change the assignment still

**Comment by atchertchian** [comment]
(also reviewed MultiTenantServiceImpl, tests showed that applicationStarted deprecated method was not called anymore, due to the start method impl, now fixed)

File: modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java (hunk 4 above)
Comment (efge) [reply modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java]: With an additional cast this seems to compile;

        service = this.<PathSegmentServiceDescriptor> getRegistryContribution(XP).map(desc -> {
            try {
                return (PathSegmentService) desc.className.getDeclaredConstructor().newInstance();
            } catch (ReflectiveOperationException e) {
                throw new RuntimeServiceException(e);
            }
        }).orElse(DEFAULT_SERVICE);

And all the lambda could be in a helper `newInstance(PathSegmentServiceDescriptor)` method to make this nicer.

File: modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/test/resources/META-INF/MANIFEST.MF [hunk 8]
```diff
@@ -0,0 +1,3 @@
+Manifest-Version: 1.0
+Bundle-Name: org.nuxeo.ecm.core.storage.sql.test
```
Comment (atchertchian) [inline modules/core/nuxeo-core-storage-sql/nuxeo-core-storage-sql/src/test/resources/META-INF/MANIFEST.MF:2]: this is making eclipse confuse this bundle and the nuxeo-core-storage-sql-test bundle: will rename it (see https://jira.nuxeo.com/browse/NXP-30098)

File: modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java (hunk 4 above)
Comment (atchertchian) [reply modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java]: I'll keep that in mind when working on https://jira.nuxeo.com/browse/NXP-30092: there are too many places where we do such instantiations (and i'd still like to throw better error messages in cases where the className would be null, which is not the case here).
The related error management should be consistent in all cases (some add runtime error messages, some not, etc...)

Comment (kevinleturc) [reply modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java]: Just for my knowledge, in such a situation, we prefer to declare the default service as a static member VS doing a `orElseGet(PathSegmentService::new)` for instance?

Comment (atchertchian) [reply modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java]: Not necessarily, at least that's not what I understood, it's just syntax-related.
Here i do agree we'd better instantiate at start, it even if it should not make a big difference (as the default class does not hold any state).
It's a bit different when this default value is non-null after the component is stopped (in ExceptionHandler component typically), where using a static member makes more sense to me, or when the instantiated class holds a state (in which case re-instantiating makes more sense).

Comment (efge) [reply modules/core/nuxeo-core-api/src/main/java/org/nuxeo/ecm/core/api/pathsegment/PathSegmentComponent.java]: Agreed that here it does not make a big difference.

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java [hunk 9]
```diff
@@ -129,4 +129,12 @@ protected String computeId(Context ctx, XAnnotatedObject xObject, Element elemen
         return (T) contrib;
     }
 
+    protected void logError(String message, Throwable t, String extensionId, String id) {
+        log.error(message, t);
+    }
+
+    protected void logWarn(String message, String extensionId, String id) {
+        log.warn(message);
+    }
```
Comment (kevinleturc) [inline modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java:138]: Logging shouldn't exist in methods.

Also, this increases the complexity as we're now having two more interfaces to provide default methods + two more classes to link the abstract and the interface.

I would prefer to use a decorator pattern for this.

Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java:138]: I don't understand the "shouldn't exist" part. Please tell if you have another implementation in mind: i started defining a "RuntimeLogger" class and thought it was overkill after all.
I can remove the whole feature for now if you prefer: the point is to easily check that we have no implicit merges by default, but it's also to prepare for consistent exception management on classes instantiation (https://jira.nuxeo.com/browse/NXP-30092)

Comment (kevinleturc) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java:138]: I wanted to say that method referring to logging shouldn't exist in code.

I really think a decorator class would be more elegant.

After a Slack talk, we agree to postpone this.

**Review by kevinleturc — APPROVED** [review]

File: modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java (hunk 9 above)
Comment (atchertchian) [reply modules/runtime/nuxeo-common/src/main/java/org/nuxeo/common/xmap/registry/MapRegistry.java]: Commit moved to the wip branch for NXP-30092

**Review by efge — APPROVED** [review]

**Pull request merged**